README.md
A library for validating national id numbers and extracting any embedded data from them.

Currenty only South Africa and Zimbabwe are implemented but more countries will be added over time

# Installation

    pip install id-validation
    
# Usage

    from id_validation import ValidatorFactory
    validator = ValidatorFactory.get_validator("ZW")

    # Use the validate method to test whether a number is valid or not according to country-specific rules
    assert validator.validate("50-025544-Q-12")

    # The extract data method returns any data that might be encoded into the id number. This is country specific.
    data = validator.extract_data("50-025544-Q-12")
    assert data["registration_region"] == "Mutasa"
    assert data["district"] == "Chivi"
    assert data["sequence_number"] == "025544"

# Countries
The following codes are available:

    BW - Botswana
    NG - Nigeria
    ZA - South Africa
    ZA_OLD - South African (Apartheid-era). See the note below for more information
    ZW - Zimbabwe


## Botswana (BW)
Note - the validation logic has been implemented from anecdotal information available online and not against official documentation.

```
>>> import id_validation
>>> from id_validation import ValidatorFactory
>>> validator = ValidatorFactory.get_validator("BW")
>>> validator.validate("379219515")
True
>>> validator.extract_data("379219515")
{'gender': 'Male'}
```

## Nigeria
Nigerian id numbers consist of 11 randomly selected digits. Find the regulations <a href="images/MandatoryNIN_Gazetted.pdf">here</a>.

```
>>> import id_validation
>>> from id_validation import ValidatorFactory
>>> validator = ValidatorFactory.get_validator("NG")
>>> validator.validate("35765421356")
True
```

## South Africa (ZA)
South African ids contain the following information:
- Date of birth
- Gender
- Citizenship (citizen or permanent resident)

```
>>> import id_validation
>>> from id_validation import ValidatorFactory
>>> validator = ValidatorFactory.get_validator("ZA")
>>> validator.validate("7106245929185")
True
>>> validator.extract_data("7106245929185")
{'dob': datetime.datetime(1971, 6, 24, 0, 0), 'gender': <GENDER.MALE: 1>, 'checksum': 5, 'citizenship': <CITIZENSHIP_TYPE.PERMANENT_RESIDENT: 1>}
```

## South Africa - Apartheid-era (ZA_OLD)
Apartheid-era South African ids contain the following information:
- Date of birth
- Gender
- Race

```
>>> import id_validation
>>> from id_validation import ValidatorFactory
>>> validator = ValidatorFactory.get_validator("ZA_OLD")
>>> validator.validate("7106245929185")
True
>>> validator.extract_data("7106245929185")
{'dob': datetime.datetime(1971, 6, 24, 0, 0), 'gender': <GENDER.MALE: 1>, 'checksum': 5, 'race': <RACE.CAPE_COLOURED: 1>}
```

### Note
These id numbers were used during the Apartheid-era. They encoded the race of the ID holder. The 1986 Identification Act removed this identifier and all id numbers were changed to the more modern version which only encodes citizenship. This validator is included for completeness. I have never seen an old id number in any dataset I have ever worked with, so avoid using it unless you are sure that your ids are pre-1986. More information can be found [here](https://web.archive.org/web/20220705233321/https://www.thoughtco.com/south-african-apartheid-era-identity-numbers-4070233)

<img src="images/apartheid_id.jpg"/>

## Zimbabwe (ZW)
Zimbabwe IDs contain the following information:
- Registration region
- Father's district

```
>>> import id_validation
>>> from id_validation import ValidatorFactory
>>> validator = ValidatorFactory.get_validator("ZW")
>>> validator.validate("50-025544-Q-12")
True
>>> validator.extract_data("50-025544-Q-12")
{'registration_region': 'Mutasa', 'district': 'Chivi', 'sequence_number': '025544'}
```


CHANGELOG
0.1 Initial Commit
0.2 Added South African and Apartheid-era South African Ids
0.3 Changed license to MIT
0.3.1 Removed unneeded requirements
0.3.2 Added __all__ to __init__.py
0.3.4 Fixed validation bug in ZimbabweValidator which did not correctly handle malformed strings
0.4.0 Added validator for Botswanan id numbers
0.4.1 Improved validation for Zimbabwean id numbers
0.4.2 Fixed South Africa apartheid id number validation
0.5.0 Added validation for Nigerian ID numbers

validate_zimbabwe.py
from .validate import ValidationError
"""
Source: Zimbabwe 2018 Elections Biometric Voters' Roll Analysis
https://www.slideshare.net/povonews/zimbabwe-2018-biometric-voters-roll-analysis-pachedu
"""
region_lookup = {
    "02": "Beitbridge",
    "03": "Mberengwa",
    "04": "Bikita",
    "05": "Bindura",
    "06": "Binga",
    "07": "Buhera",
    "08": "Bulawayo",
    "10": "Unknown", # (Undetermined) (believed: Mhondoro- "11": "Ngezi)    ",
    "11": "Muzarabani",
    "12": "Chivi",
    "13": "Chipinge",
    "14": "Chiredzi",
    "15": "Mazowe",
    "18": "Chikomba",
    "19": "Umzingwane",
    "21": "Insiza",
    "22": "Masvingo",
    "23": "Gokwe South",
    "24": "Kadoma",
    "25": "Goromonzi",
    "26": "Gokwe North",
    "27": "Gutu",
    "28": "Gwanda",
    "29": "Gweru",
    "32": "Chegutu",
    "34": "Nyanga",
    "35": "Bubi",
    "37": "Kariba",
    "38": "Hurungwe",
    "39": "Matobo",
    "41": "Lupane",
    "42": "Makoni",
    "43": "Marondera",
    "44": "Chimanimani",
    "45": "Mt. Darwin",
    "46": "Unknown", # (Undetermined) (believed: Mbire)
    "47": "Murehwa",
    "48": "Mutoko",
    "49": "Mudzi",
    "50": "Mutasa",
    "53": "Nkayi",
    "54": "Mwenezi",
    "56": "Bulilimamangwe",
    "58": "Kwekwe",
    "59": "Seke",
    "61": "Rushinga",
    "63": "Harare",
    "66": "Shurugwi",
    "67": "Zvishavane",
    "68": "Shamva",
    "70": "Makonde",
    "71": "Guruve",
    "73": "Tsholotsho",
    "75": "Mutare",
    "77": "Chirumanzu",
    "79": "Hwange",
    "80": "Hwedza",
    "83": "Zaka",
    "84": "Umguza",
    "85": "U.M.P. (Uzumba, Maramba, Pfungwe)",
    "86": "Zvimba",
}

class ZimbabweValidator:
    def _get_region(self, region_id: str) -> str:
        """
        Returns the region name for the given region id
        """
        try:
            return region_lookup[region_id]
        except IndexError:
            raise ValidationError(f"Invalid region code: {region_id}")

    def _checksum(self, id_number: str) -> bool:
        """
        Validates the checksum digit of the given id number using mod 23
        """
        registration_code, sequence_number, check_letter, _ = self._extract_parts(id_number)
        check_number = int(registration_code + sequence_number)
        mod = check_number % 23
        lookup = {
            1: "A", 2: "B", 3: "C", 4: "D",
            5: "E", 6: "F", 7: "G", 8: "H",
            9: "J", 10: "K", 11: "L", 12: "M",
            13: "N", 14: "P", 15: "Q", 16: "R",
            17: "S", 18: "T", 19: "V", 20: "W",
            21: "X", 22: "Y", 23: "Z",
        }

        return check_letter == lookup[mod]

    def _clean_id_number(self, id_number: str) -> str:
        return id_number.replace("-", "").replace(" ", "")

    def _extract_parts(self, id_number: str) -> list[str]:
        registration_code = id_number[0:2]
        sequence_number = id_number[2:-3]
        check_letter = id_number[-3]
        district_code = id_number[-2:]

        return registration_code, sequence_number, check_letter, district_code


    def validate(self, id_number: str) -> bool:
        """
        Verify the region codes as well as use the modulus 23 validation
        """
        cleaned_id_number = self._clean_id_number(id_number)
        registration_code, _, _, district_code = self._extract_parts(cleaned_id_number)

        try:
            _ = self._get_region(registration_code)
            _ = self._get_region(district_code)
        except ValidationError:
            return False

        return self._checksum(cleaned_id_number)

    def extract_data(self, id_number: str) -> dict[str, str]:
        is_valid = self.validate(id_number)
        if not is_valid:
            raise ValidationError("Invalid ID number")
        
        cleaned_id_number = self._clean_id_number(id_number)
        registration_code, sequence_number, _, district_code = self._extract_parts(cleaned_id_number)

        return {
            "registration_region": self._get_region(registration_code),
            "district": self._get_region(district_code),
            "sequence_number": sequence_number
        }



